src/Microsoft.AspNet.Routing/VirtualPathData.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Routing
+{
+    /// <summary>
+    /// Represents information about the route and virtual path that are the result of
+    /// generating a URL with the ASP.NET routing middleware.
+    /// </summary>
+    public class VirtualPathData
+    {
+        private string _virtualPath;
+        private readonly IDictionary<string, object> _dataToken;
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="VirtualPathData"/> class.
+        /// </summary>
+        /// <param name="router">The object that is used to generate the URL.</param>
+        /// <param name="virtualPath">The generated URL.</param>
+        public VirtualPathData([NotNull] IRouter router, string virtualPath)
+            : this(router, virtualPath, dataTokens: new RouteValueDictionary())
+        {
+        }
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="VirtualPathData"/> class.
+        /// </summary>
+        /// <param name="router">The object that is used to generate the URL.</param>
+        /// <param name="virtualPath">The generated URL.</param>
+        /// <param name="dataTokens">The collection of custom values.</param>
+        public VirtualPathData(
+            [NotNull] IRouter router,
+            string virtualPath,
+            IDictionary<string, object> dataTokens)
+        {
+            _dataToken = dataTokens ?? new RouteValueDictionary();
```

Comment:
The VirtualPathData is a snapshot so that it can be mutated without modifying the original data. Remember, the goal here is to preserve behavior from System.Web's Routing code.